Add toggle for favorites in create_root_dialog

willnationsdev opened the pull request:
This is in reference to @akien-mga and @groud's suggestion at the end of #20238.

It creates an alternative scene root creation dialog that can be toggled in the EditorSettings. It enables users to have their favorites list show up in place of the beginner-friendly `create_root_dialog`. It also supports the creation of script classes from this list.

It all updates in real-time as things are changed.

Toggled Off:
![toggle_off](https://user-images.githubusercontent.com/16217563/43284530-b5a2b504-90e1-11e8-8b92-9e1cfa655faf.png)

Toggled On:
![toggle_on](https://user-images.githubusercontent.com/16217563/43284533-b9562bd6-90e1-11e8-9f82-f1f65139b23b.png)

Unfavorited Node:
![unfavorited](https://user-images.githubusercontent.com/16217563/43284556-caffcb44-90e1-11e8-9862-f09ddc9e9c41.png)

Favorited Node:
![favorited](https://user-images.githubusercontent.com/16217563/43284563-d042d3b2-90e1-11e8-967b-2af4c48ea9ea.png)

EDIT:

The change is now toggled on or off via signal callbacks from an in-panel button rather than making the process logic more complicated. Here is what it looks like now:

Button Toggled Off:
![button_off](https://user-images.githubusercontent.com/16217563/43734072-33b3aef6-997c-11e8-9ad9-5039b1dc3b90.png)

Button Toggled On:
![button_on](https://user-images.githubusercontent.com/16217563/43734081-38142d90-997c-11e8-8dfe-1664fbcc1d43.png)

If you have a script class and the script class's information is updated while it is in this list, it will be automatically removed from the list (since it ceased to exist). The serialized favorites data should automatically be updated too.

willnationsdev committed:
Add toggle for favorites in create_root_dialog.

swarnimarun commented:
At first I didn't understand what this PR was but man this is awesome. 
But won't it be better to be able to save entire scene trees as templates??

It could be an additional feature but it will be amazing to have I think.

willnationsdev commented:
> save entire scene trees as templates

That's effectively the same thing as inheriting a scene, which is already a feature.

btw, I'm currently re-working this PR so that it doesn't even create a separate dialog (just updates the existing one).

swarnimarun commented:
I am not saying that it's not possible just want to make the accessibility greater.

Finally Visual Scripting has got some much needed changes so I think we will see more beginners devs flock here and make somethings easier.

willnationsdev commented:
@swarnimarun 

Well, something that might(?) happen eventually is being able to add scenes to the CreateDialog. And if that were to ever happen, then it wouldn't be a stretch to add such scenes to your Favorites (thereby making them show up in this list).

I mentioned those negations earlier because that's the general criteria for adding things to Godot. You are right though, being able to add scenes to a quick-build list when making a new scene would be extremely useful.

willnationsdev commented:
The PR has now been updated to not use a separate dialog (much simpler implementation).

akien-mga commented:
Screenshots/gif would be nice :)

willnationsdev commented:
@akien-mga Updated. :-)

akien-mga commented:
I like the feature, but I'm not convinced by the way it's implemented, it seems a bit convoluted to remove all the init from the `_ready()` and instead put it in a update method that will run in process, potentially recreating all the scene at each iteration, but actually it won't because a cached variable will prevent it... Sounds overly complicated. I'm not sure what a better design would be, but from a quick glance I'm not convinced by the current one.

willnationsdev commented:
I updated it with a bug fix that makes it so that the favorites list is re-serialized/updated anytime the script classes are updated. This stops the favorites list from getting clogged with previously-favorited, now-deleted script classes.

willnationsdev commented:
Well, the "run in process but don't do anything if this boolean is set" concept was there from the beginning. I just changed the boolean to be a combination of values that are all meant to detect (has the state changed). If there were a way to trigger it exclusively with signals, maybe that'd be better...

willnationsdev commented:
@akien-mga Looks like there's a "settings_changed" signal in EditorSettings I could connect to. Would need to find one for changing scene tabs too (so if someone creates a new scene tab that would be empty) and also 1 or 2 signals for creating/deleting nodes in the scene tree. I mean, that could be a lot messier based on the sheer number of signals that would have to be managed to make it work.

akien-mga commented:
I don't mean process shouldn't be used to see when the scene gets/has a root node, but the _update* method should not be the one responsible for instancing the panel with new memory allocations IMO. This should be done once, and hidden when not needed/updated when favorites change.

akien-mga commented:
But yeah I only had a quick look, maybe it's good enough as is :)

willnationsdev commented:
@akien-mga Yeah, both if those operations (show/hide, update when favorites change, and also update when script classes change) all happen during that method, which is only called when it is necessary.

scayze commented:
Wouldn't it make sense to have this in the project settings instead? Especially now that you have script classes, which are project specific?

groud commented:
> Wouldn't it make sense to have this in the project settings instead? Especially now that you have script classes, which are project specific?

i don't think so, when you're experienced enough to enable this feature you want it for all your project.
And it's better this way in terms of discoverability

willnationsdev commented:
@akien-mga I like that idea. Should be able to implement that easily enough, and it would probably simplify the overall implementation a lot too since I could use the signal of the toggle button to trigger the swap.

willnationsdev commented:
I've updated this PR to rely more strongly on signal callbacks and minimal node updates. Let me know what you think.

willnationsdev commented:
I've updated the OP with new pictures that show the button in-action.

willnationsdev commented:
Fixed a bug where the EditorSettings modification was still visible to the user from the EditorSettings editing dialog (it should now be a serialized to a key that is not user-editable).

akien-mga commented:
Thanks!
